exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -448,39 +427,44 @@ public Builder withValidators(short validatorCount) {
     }
 
     /**
-     * Adds a service with which the TestKit would be instantiated. Several services can be added.
+     * Adds a service artifact which would be deployed by the TestKit. Several service artifacts
+     * can be added.
+     *
+     * <p>Note that the corresponding service instance with equal serviceArtifactId should be
+     * created with {@link #withService(ServiceArtifactId, String, int, Any)}.
      */
-    public Builder withService(Class<? extends ServiceModule> serviceModule) {
-      services.add(serviceModule);
+    public Builder withDeployedService(ServiceArtifactId serviceArtifactId, String artifactFilename) {
+      serviceArtifactFilenames.put(serviceArtifactId, artifactFilename);
       return this;
     }
 
     /**
-     * Adds services with which the TestKit would be instantiated.
-     */
-    @SafeVarargs
-    public final Builder withServices(Class<? extends ServiceModule> serviceModule,
-                                      Class<? extends ServiceModule>... serviceModules) {
-      return withServices(Lists.asList(serviceModule, serviceModules));
-    }
-
-    /**
-     * Adds services with which the TestKit would be instantiated.
+     * Adds a service specification with which the TestKit would create the corresponding service
+     * instance. Several service specifications can be added.
+     *
+     * <p>Note that the corresponding service artifact with equal serviceArtifactId should be
+     * deployed with {@link #withDeployedService(ServiceArtifactId, String)}.
      */
-    public Builder withServices(Iterable<Class<? extends ServiceModule>> serviceModules) {
-      Iterables.addAll(services, serviceModules);
+    public Builder withService(ServiceArtifactId serviceArtifactId,
+                               String serviceName,
+                               int serviceId,
+                               Any configuration) {
+      ServiceSpec serviceSpec =
+          ServiceSpec.newInstance(serviceName, serviceId, configuration.toByteArray());
+      services.put(serviceArtifactId, serviceSpec);
       return this;
     }
 
     /**
-     * If called, will create a TestKit with time service enabled. The time service will use the
-     * given {@linkplain TimeProvider} as a time source.
+     * If called, will create a TestKit with time service enabled. The time service will be created
```

Comment:
Please rephrase as `withService`, as with multiple instances supported they must work the same.